r/src/array_from_vector.cpp

```diff
@@ -797,84 +909,118 @@ Status GetConverter(const std::shared_ptr<DataType>& type,
   return Status::NotImplemented("type not implemented");
 }
 
-template <typename Type>
-std::shared_ptr<arrow::DataType> GetFactorTypeImpl(bool ordered) {
-  return dictionary(std::make_shared<Type>(), arrow::utf8(), ordered);
+static inline std::shared_ptr<arrow::DataType> IndexTypeForFactors(int n_factors) {
+  if (n_factors < INT8_MAX) {
+    return arrow::int8();
+  } else if (n_factors < INT16_MAX) {
+    return arrow::int16();
+  } else {
+    return arrow::int32();
```

Comment:
Since the R storage is int32, would it be better to always use int32 (so we could zero-copy the indices)?

Reply:
Good point.